mmdet/models/detectors/base_detr.py

```diff
@@ -0,0 +1,123 @@
+            data_preprocessor=data_preprocessor, init_cfg=init_cfg)
+        # process args
+        bbox_head.update(train_cfg=train_cfg)
+        bbox_head.update(test_cfg=test_cfg)
+        self.train_cfg = train_cfg
+        self.test_cfg = test_cfg
+        self.encoder_cfg = encoder_cfg
+        self.decoder_cfg = decoder_cfg
+        self.positional_encoding_cfg = positional_encoding_cfg
+        self.num_query = num_query
+
+        # init model layers
+        self.backbone = MODELS.build(backbone)
+        if neck is not None:
+            self.neck = MODELS.build(neck)
+        self.bbox_head = MODELS.build(bbox_head)
+        self._init_layers()
+
+    @abstractmethod
+    def _init_layers(self) -> None:
+        """Initialize layers except backbone, neck and bbox_head."""
+        pass
+
+    def loss(self, batch_inputs: Tensor,
+             batch_data_samples: SampleList) -> Union[dict, list]:
+        img_feats = self.extract_feat(batch_inputs)
+        transformer_inputs_dict = self.forward_pretransformer(
+            img_feats, batch_data_samples)
+        outs_dec = self.forward_transformer(**transformer_inputs_dict)
+        losses = self.bbox_head.loss(outs_dec, batch_data_samples)
+        return losses
+
+    def predict(self,
+                batch_inputs: Tensor,
+                batch_data_samples: SampleList,
+                rescale: bool = True) -> SampleList:
+        img_feats = self.extract_feat(batch_inputs)
+        transformer_inputs_dict = self.forward_pretransformer(
+            img_feats, batch_data_samples)
+        outs_dec = self.forward_transformer(**transformer_inputs_dict)
+        results_list = self.bbox_head.predict(
+            outs_dec, batch_data_samples, rescale=rescale)
+        batch_data_samples = self.add_pred_to_datasample(
+            batch_data_samples, results_list)
+        return batch_data_samples
+
+    def _forward(
+            self,
+            batch_inputs: Tensor,
+            batch_data_samples: OptSampleList = None) -> Tuple[List[Tensor]]:
+        """Network forward process.
+
+        Includes backbone, neck and head forward without post-processing.
+        """
+        img_feats = self.extract_feat(batch_inputs)
+        transformer_inputs_dict = self.forward_pretransformer(
+            img_feats, batch_data_samples)
+        outs_dec = self.forward_transformer(**transformer_inputs_dict)
+        results = self.bbox_head.forward(outs_dec)
+        return results
+
+    def extract_feat(self, batch_inputs: Tensor) -> Tuple[Tensor]:
+        """Extract features.
+
+        Args:
+            batch_inputs (Tensor): Image tensor with shape (N, C, H ,W).
+
+        Returns:
+            tuple[Tensor]: Multi-level features that may have
+            different resolutions.
+        """
+        x = self.backbone(batch_inputs)
+        if self.with_neck:
+            x = self.neck(x)
+        return x
+
+    @abstractmethod
+    def forward_pretransformer(
```

Comment:
Add a general comment
` This function creates the inputs of the Transformer`

Reply:
Resolved.